ui-v2/app/controllers/dc/acls/policies/edit.js

```diff
@@ -0,0 +1,45 @@
+import Controller from '@ember/controller';
+import { inject as service } from '@ember/service';
+import { get, set } from '@ember/object';
+export default Controller.extend({
+  builder: service('form'),
+  dom: service('dom'),
+  isScoped: false,
+  init: function() {
+    this._super(...arguments);
+    this.form = get(this, 'builder').form('policy');
+  },
+  setProperties: function(model) {
+    // essentially this replaces the data with changesets
+    this._super(
+      Object.keys(model).reduce((prev, key, i) => {
+        switch (key) {
+          case 'item':
+            prev[key] = this.form.setData(prev[key]).getData();
+            break;
+        }
+        return prev;
+      }, model)
+    );
+    set(this, 'isScoped', get(model.item, 'Datacenters.length') > 0);
```

Comment:
> I liken it to overriding an Array or String prototype. Sure Array#push is a public method but anyone familiar with JavaScript has a preconceived idea of what it does

I'd definitely not recommend altering any native objects in any way without extending them first. I think we've spoken about this before here https://github.com/hashicorp/nomad/pull/4635#discussion_r217655431

I've just looked again at the file, aren't you still overriding Array instance methods here? 

https://github.com/hashicorp/nomad/blob/adc05976c2c7f97d46c214518d448b0b13e377d6/ui/app/utils/classes/rolling-array.js

It's not the prototype method, but you couldn't you still do an

```javascript
array.constructor === Array ? array.push(thing) : somethingElse
```

 ...and end up with something you didn't expect?

In my case I'm _extending_ `Controller` first, so I'm essentially adding functionality to a base `Controller` to give me a specific `PoliciesController` which is my own custom class. `setProperties` is public and not final in anyway (AFAIK) so usually that would mean 'fair game', as long as it does what you expect aswell as the extra functionality you add (which it does by calling `super`). Additionally what my custom `setProperties` does is to 'proxy' certain variables you pass it (which I've mentioned here https://github.com/hashicorp/consul/pull/4789/files/88027a940b088d4e3c96b8dbcd9e557c8d524881#r225548050). I also left a comment in to explain this slightly:

https://github.com/hashicorp/consul/blob/88027a940b088d4e3c96b8dbcd9e557c8d524881/ui-v2/app/controllers/dc/acls/policies/edit.js#L13

Changesets are just proxy objects to the underlying data. So it just 'decorates' the data with extra functionality.

> Something like controller.setProperties(controller.syncFormData({ prop, er, ties })) would be more obvious upon reading.

One thing I should explain here, is that I went to great lengths to keep the `forms` in the View/ViewController layer, and out of the Routes. They are a view concern, so this is where they should live.

I can appreciate your overall point a little bit and I did think of other options and I've thought of others to try to address this somehow. The answer to the problem here is to find a hook to use in the View layer immediately before the View is rendered. I've considered `setProperties` to be this hook as the View is re-rendered immediately after that.

Other things I can think of:

1. Use a computed property: 
```javascript
actualItem: computed('item', function(){
  return this.form.setData()
})
```
This now means that `item` in my template is no longer the `item` that I want. It's the unproxied `item`. To me this is surprising as I'm passing it into the `View` as `item`, if I look at the templates it leaves me thinking "what's `actualItem`, what happened to `item` from my Route?" Plus I'd have to do this for every property I want to proxy, which would get tedious very quickly. Right now I can just loop over the props in `setProperties` and proxy the object there.

2. Use computed setter/getters:
```javascript
item: computed('item', function(){
  {
    set: function(){},
    get: function(){}
  }
})
```

As far as I know this means having yet another property on the Controller to save the actual value of `item`, I'd also have to do this for every property (see above).

3. Use a more generically named controller 'setup' function, and remember to call it everytime I call `setProperties`.
```javascript
export Route.extend({
  setupController(controller, model) {
   
    controller.setProperties({
      some: 'things',
      go: here,
    });
    if(typeof controller.setup === 'function') {
      controller.setup();
    }
    return this._super(...arguments);
  },
});
```
`controller.setup` would mutate the `item` property into the proxy object. But it's named something generic enough to make sense. I'm actually surprised that I can't find any default `setup/reset`-like methods for `Controller` in `ember`.

Option 3 is my favourite at a pinch, but there is always the danger that you'd forget to call `setup` somewhere, and therefore produce some hard to spot bugs.

There's even a plugin which attempts to make this less error prone for you:

https://github.com/kellyselden/ember-controller-lifecycle

The only problem with this is, there is _still_ a danger that you might call `setProperties` somewhere else and your `controller.setup()` is never called, which would lead to unexpected bugs.

Do you have any further ideas following the above info?

(Apologies if this comment is a bit lengthy, I want to try to make sure follow my thinking here)

Reply:
Sorry had another idea for this, thought I'd pop it in here incase.

> Option 3 is my favourite at a pinch, but there is always the danger that you'd forget to call setup somewhere, and therefore produce some hard to spot bugs.

Option 4:

If I changed things so that I stopped calling `setProperties`, but instead called a method on the `Controller` that _then_ called `setProperties`. So something like:

```javascript
// ReopenedController using a similar approach to ember-controller-lifecycle
newMethodNameThatsGenericEnough: function(model) {
  const proxiedModel = this.doExtraStuffWithModel();
  this.setProperties(proxiedModel);
}

// SomeRoute
setupController: function(controller, model) {
  this._super(...arguments);
  this.newMethodNameThatsGenericEnough(model);
}
```

..and now when I have anywhere in my codebase where I was previously calling `setProperties` and there was the danger of forgetting to then call a `setup`, I just change it to call `newMethodNameThatsGenericEnough` instead.

I don't think this method should be called `setup` though as I also call `setProperties` when I'm not setting things up, so it wouldn't make sense there. Will think a bit more on a good name. But I think this addresses my problem and satisfies your concerns about overwriting the `setProperties` method, so win win.

The only tiny thing is, this isn't 'conventional `ember`' we've added a new undocumented method - but I'm ok with that if you are?

Reply:
Related: https://github.com/hashicorp/consul/pull/5056